src/ucp/proto/proto_am.c

```diff
@@ -114,3 +114,30 @@ void ucp_proto_am_zcopy_completion(uct_completion_t *self)
 
     ucp_proto_am_zcopy_req_complete(req, self->status);
 }
+
+ucs_status_t ucp_proto_am_req_copy_header(ucp_request_t *req)
+{
+    void *user_header;
+
+    if ((req->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED) ||
+        (req->send.msg_proto.am.header.length == 0)) {
+        return UCS_OK;
+    }
+
+    if (req->send.msg_proto.am.flags & UCP_AM_SEND_FLAG_COPY_HEADER) {
+        user_header =
+                ucs_mpool_set_get_inline(&req->send.ep->worker->am_mps,
+                                         req->send.msg_proto.am.header.length);
+        if (ucs_unlikely(user_header == NULL)) {
+            ucs_error("failed to allocate memory");
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        memcpy(user_header, req->send.msg_proto.am.header.user_ptr,
+               req->send.msg_proto.am.header.length);
+        req->flags |= UCP_REQUEST_FLAG_USER_HEADER_COPIED;
+        req->send.msg_proto.am.header.user_ptr = user_header;
```

Comment:
Maybe we should consider a different name to replace "user_ptr" because after copy the header it isn't user ptr anymore.
Options:
- desc
- data
- ptr
- buffer

Reply:
maybe add a union like this
```
union {
    user_ptr;
    copy;
};
```
and access them according to `UCP_REQUEST_FLAG_USER_HEADER_COPIED`?

Reply:
Good idea. I don't see any reason why it should not work. 
Will go with this approach.

Reply:
I meant that accessing them according to UCP_REQUEST_FLAG_USER_HEADER_COPIED will add 
unnecessary branches to the pack functions (to determine the header pointer to copy from).

Reply:
I think compiler is able to recognize it as well and compile out the branches.

Reply:
Per offline discussion:
- rename req->send.msg_proto.am.header.user_ptr to req->send.msg_proto.am.header.**ptr**
- @evgeny-leksikov to implement clean for am_zcopy that release zcopy desc and also remove the check https://github.com/openucx/ucx/pull/8385/files#diff-fbdbc6d113f59edb905877378b7d64e4931e407f5c144f22dd5ebf70ccb4c5beR62
- Zcopy cleanup (but only if user asked for persistent header) will copy header from reg_desc to new desc in header.ptr - since user considers his original pointer as reusable
- All protocol completion functions should check UCP_REQUEST_FLAG_USER_HEADER_COPIED and release **ptr** accordingly to mpool. Even zcopy, which does not copy the header itself, may start with a request where header was already copied by prev protocol.
- AM zcopy protocol just copies header.ptr to zcopy desc, but does not invalidate header.ptr